hydromt/components/base.py

```diff
@@ -0,0 +1,51 @@
+"""Provides the base class for model components."""
+
+from abc import ABC, abstractmethod
+from logging import Logger
+from typing import TYPE_CHECKING, cast
+from weakref import ReferenceType, ref
+
+from hydromt.data_catalog import DataCatalog
+
+if TYPE_CHECKING:
+    from hydromt.models import Model
+    from hydromt.root import ModelRoot
+
+
+class ModelComponent(ABC):
+    """Abstract base class for ModelComponent."""
+
+    def __init__(self, model: "Model"):
+        self.__model_ref: ReferenceType["Model"] = ref(model)
+
+    @abstractmethod
+    def read(self):
+        """Read the file(s) into the component."""
+        if not self._model_root.is_reading_mode():
+            raise IOError("Model opened in write-only mode")
+
+    @abstractmethod
+    def write(self):
+        """Write the component to file(s)."""
+        if not self._model_root.is_writing_mode():
+            raise IOError("Model opened in read-only mode")
+
+    @property
+    def _model(self) -> "Model":
+        """Return the model object this component is associated with."""
+        return cast("Model", self.__model_ref())
+
+    @property
+    def _data_catalog(self) -> DataCatalog:
+        """Return the data catalog of the model this component is associated with."""
+        return self._model.data_catalog
+
+    @property
+    def _logger(self) -> Logger:
+        """Return the logger of the model this component is associated with."""
+        return self._model.logger
+
+    @property
+    def _model_root(self) -> "ModelRoot":
```

Comment:
We encourage component developers to use these properties, as they are utility properties for not writing `self._model.logger` etc. so often, but we don't want to expose them to a user. All this information can already be found under `Model`, and otherwise we would expose things twice, making the API overflowing with functions that users don't need to see.

```py
# Some user code
m = Model()
grid = GridComponent(m)
m.add_component("grid", grid)

dc = m.data_catalog # makes sense, it is the data catalog of the whole model
dc2 = m.grid.data_catalog # does not make sense to me. It would be a superfluous function
```